catalyst/callbacks/lambda_preprocess.py

```diff
@@ -0,0 +1,184 @@
+            input_keys = [input_keys]
+        if output_keys is not None:
+            if not isinstance(output_keys, (list, str)):
+                raise TypeError("output keys should be str or list of str.")
+            if isinstance(output_keys, str):
+                output_keys = [output_keys]
+        else:
+            output_keys = input_keys
+        self.keys_to_apply = input_keys
+        self.output_keys = output_keys
+        self.lambda_fn = lambda_fn
+        self.handler: Union[None, Callable] = None
+
+    def on_batch_end(self, runner) -> None:
+        """
+        On batch end action.
+
+        Args:
+            runner: runner for the experiment.
+
+        Raises:
+            TypeError: If lambda_fn output has a wrong type.
+
+        """
+        fn_inp = [runner.batch[key] for key in self.keys_to_apply]
+        fn_output = self.lambda_fn(*fn_inp)
+        if self.handler is None:
+            if isinstance(fn_output, dict):
+                self.handler = self._handle_output_dict
+            if isinstance(fn_output, tuple):
+                if len(fn_output) != len(self.output_keys):
+                    raise TypeError(
+                        "Unexpected output. "
+                        "Expect function to return tuple same length as output_keys, "
+                        f"which is {len(self.output_keys)}, "
+                        f"but got output length of {len(fn_output)}"
+                        "Use output_keys argument to specify output keys."
+                    )
+                self.handler = self._handle_output_tuple
+            else:
+                if len(self.output_keys) > 1:
+                    raise TypeError(
+                        "Unexpected output. "
+                        "Expect function to return tuple or dict, but got "
+                        f"{type(fn_output)}. "
+                        "Use output_keys argument to specify output key."
+                    )
+                self.handler = self._handle_output_value
+        runner.batch = self.handler(
+            batch=runner.batch, function_output=fn_output, output_keys=self.output_keys
+        )
+
+    @staticmethod
+    def _handle_output_tuple(
+        batch: Dict[str, Any], function_output: Tuple[Any], output_keys: List[str]
+    ) -> Dict[str, Any]:
+        for out_idx, output_key in enumerate(output_keys):
+            batch[output_key] = function_output[out_idx]
+        return batch
+
+    @staticmethod
+    def _handle_output_dict(
+        batch: Dict[str, Any], function_output: Dict[str, Any], output_keys: List[str]
+    ) -> Dict[str, Any]:
+        for output_key, output_value in function_output.items():
+            batch[output_key] = output_value
+        return batch
+
+    @staticmethod
+    def _handle_output_value(
+        batch: Dict[str, Any], function_output: Any, output_keys: List[str],
+    ):
+        batch[output_keys[0]] = function_output
+        return batch
```

Comment:
could we place them above the `on_batch_end` so it should be seen before the use case?